src/contracts/core/DelegationManager.sol

```diff
@@ -412,46 +503,217 @@ contract DelegationManager is Initializable, OwnableUpgradeable, Pausable, Deleg
+            "DelegationManager.completeQueuedAction: only withdrawer can complete action"
+        );
+
+        if (receiveAsTokens) {
+            require(
+                tokens.length == withdrawal.strategies.length, 
+                "DelegationManager.completeQueuedAction: input length mismatch"
+            );
+        }
+
+        // Remove `withdrawalRoot` from pending roots
+        delete pendingWithdrawals[withdrawalRoot];
+
+        address currentOperator = delegatedTo[msg.sender];
+
+        // Finalize action by converting shares to tokens for each strategy, or
+        // by re-awarding shares in each strategy.
+        for (uint256 i = 0; i < withdrawal.strategies.length; ) {
+            if (receiveAsTokens) {
+                _withdrawSharesAsTokens({
+                    staker: withdrawal.staker,
+                    withdrawer: msg.sender,
+                    strategy: withdrawal.strategies[i],
+                    shares: withdrawal.shares[i],
+                    token: tokens[i]
+                });
+            // Award shares back in StrategyManager/EigenPodManager. If withdrawer is delegated, increase the shares delegated to the operator
+            } else {
+                /** When awarding podOwnerShares in EigenPodManager, we need to be sure to only give them back to the original podOwner.
+                 * Other strategy sharescan + will be awarded to the withdrawer.
+                 */
+                if (withdrawal.strategies[i] == beaconChainETHStrategy) {
+                    address staker = withdrawal.staker;
+                    /**
+                    * Update shares amount depending upon the returned value.
+                    * The return value will be lower than the input value in the case where the staker has an existing share deficit
+                    */
+                    uint256 increaseInDelegateableShares = eigenPodManager.addShares({
+                        podOwner: staker,
+                        shares: withdrawal.shares[i]
+                    });
+                    address podOwnerOperator = delegatedTo[staker];
+                    // Similar to `isDelegated` logic
+                    if (podOwnerOperator != address(0)) {
+                        _increaseOperatorShares({
+                            operator: podOwnerOperator,
+                            // the 'staker' here is the address receiving new shares
+                            staker: staker,
+                            strategy: withdrawal.strategies[i],
+                            shares: increaseInDelegateableShares
+                        });
+                    }
+                } else {
+                    strategyManager.addShares(msg.sender, withdrawal.strategies[i], withdrawal.shares[i]);
+                    // Similar to `isDelegated` logic
+                    if (currentOperator != address(0)) {
+                        _increaseOperatorShares({
```

Comment:
Nit: could we read and assign this variable just above, like in the EPmgr branch? Currently we're grabbing `currentOperator` from way earlier in the method, and it's only used here.

Would also save a storage read for the EPmgr branch!

Reply:
I added this commit https://github.com/Layr-Labs/eigenlayer-contracts/pull/222/commits/f0ecb9272b8b87c99c7154566a6feef02c9b76de
We still end up SLOAD'ing in the EPM branch, but I think it's fine.
The code is a little awkward with the two ~identical loop definitions, but it saves an SLOAD in some cases + helps make the semantic meaning of `receiveAsTokens` clearer (IMO, at least).
Appreciate the comment.

Reply:
yup yup. We could revert the commit, but I prefer the improvement in semantics with the change, especially w/r/t `receiveAsTokens`. Planning to leave this as-is.